feat(thermocycler-gen2): overshoot/undershoot tuning

fsinapi opened the pull request:
This PR makes some big changes to how the overshoot/undershoot behavior is handled on the Thermocycler Gen2.

* Default Kp has been increased quite a bit, results still have minimal unwanted overshoot/oscillation
* Added a factor of 2x when looking for the Proportional Band to allow slightly more integrator windup before the P term starts to damp
* The amplitude of the goal overshoot/undershoot is now 2º for every 50µL of liquid
* Instead of a 10-second "hold period," the TC2 switches its target to the user-programmed temperature as soon as the overshoot/undershoot target is reached.
  * There is a 1º margin here, which is accounted for by increasing the overshoot/undershoot magnitude by a flat 1º. This ensures that the PID controller errs on the side of aggressive ramping to the temperature, rather than requiring a separate tune.
  * The center channel gets an extra boost after testing showed that the temperature in the wells responds slower than other wells.
  * The block is only considered "at" this target temperature once all three columns independently reach their overshoot/undershoot target
* Overshoot/undershoot is **not** applied in the following circumstances:
  * Sample volume is less than 20µL
  * Current temperature is less than 5º away from the target temperature
  * Target temperature is below ambient room temperature
* Hold time countdown is active when the following conditions are met:
  * The plate temperature has reached the 2º threshold for the overshoot/undershoot, and is now in the "steady state" phase targeting the programmed target
  * The plate temperature is within 2º of the programmed target temperature

Merging this PR pends on some test data using an external temperature probe.

Some captures of test data with various programmed volumes:
![image](https://user-images.githubusercontent.com/27798632/186262169-6179111d-c45e-461f-84e4-4e51dc4d25e8.png)

![image](https://user-images.githubusercontent.com/27798632/186262250-1808089e-0803-4ef2-bb93-d0f850376f2e.png)

![image](https://user-images.githubusercontent.com/27798632/186262302-8289afad-463d-4ca6-9841-5db7ae6ede7f.png)

fsinapi committed:
Updated overshoot/undershoot calculation scaling and minimum difference to trigger it

fsinapi committed:
* Do not activate overshoot/undershoot if the target temperature is <5º away, if the hold time is >2 minutes, or if the target is lower than room temp
* Increased default KP to favor somewhat more aggressive transients
* Increased effective proportional band by 2x to reduce weird ramps when approaching target temperature
* Start counting hold time when within 2 degrees of target
* Move on from overshoot phase as soon as the temperature is within 2 degrees of the overshoot temp

fsinapi committed:
Updated `test_pcr_cycle` to use the thermocycler's hold time tracking
* Updated hard-coded constants to constexpr variables
* Fixed plate control tests to work with new code

fsinapi committed:
Merge branch 'edge' into RET-1183_overshoot_scaling

sfoster1 reviewed (approved):
This looks good to me, nice graphs!

fsinapi committed:
Merge branch 'edge' into RET-1183_overshoot_scaling

fsinapi committed:
Added extra overshoot to middle channel
- Don't move on from overshoot step until every channel is at the overshoot temp
- Updated I term to 0.05

fsinapi committed:
Fixed linter errors

fsinapi committed:
Fixed test for plate control

sfoster1 reviewed (approved):
Changes look good to me.
